src/ipc/ipc-common.c

```diff
@@ -117,18 +117,44 @@ struct ipc_comp_dev *ipc_get_ppl_comp(struct ipc *ipc, uint32_t pipeline_id, int
 		/* first try to find the module in the pipeline */
 		if (dev_comp_pipe_id(icd->cd) == pipeline_id) {
 			struct list_item *buffer_list = comp_buffer_list(icd->cd, dir);
+			bool b_endpoint = true;
 
 			/* The component has no buffer in the given direction */
 			if (list_is_empty(buffer_list))
 				return icd;
 
-			/* it's connected pipeline, so find the connected module */
-			buffer = buffer_from_list(buffer_list->next, struct comp_buffer, dir);
-			buff_comp = buffer_get_comp(buffer, dir);
-
-			/* Next component is placed on another pipeline */
-			if (buff_comp && dev_comp_pipe_id(buff_comp) != pipeline_id)
+			/* In the following pipelines, component.pipeline2.1 and
+			 * component.pipeline2.2 will be identified as endpoint since they are
+			 * connected to different pipelines, but only component.pipeline2.1 is
+			 * the endpoint of pipeline2 since the one source of component.pipeline2.2
+			 * is still in the same pipeline.
+			 *
+			 *                            component.pipeline1.0
+			 *                                      |
+			 *                                     \ /
+			 *                            component.pipeline2.1
+			 *                                      |
+			 *                                     \ /
+			 *
+			 *   component.pipeline3.0 -> component.pipeline2.2
+			 *                                      |
+			 *                                     \ /
+			 *                            component.pipeline2.3
+			 */
+			list_for_item(item, buffer_list) {
+				buffer = buffer_from_list(item, struct comp_buffer, dir);
+				buff_comp = buffer_get_comp(buffer, dir);
+
+				if (buff_comp && dev_comp_pipe_id(buff_comp) == pipeline_id) {
+					b_endpoint = false;
+					break;
+				}
+			}
+
+			if (b_endpoint) {
 				next_ppl_icd = icd;
+				break;
```

Comment:
can we just `return icd;` here and `return NULL;` below and remove `next_ppl_icd`?

Reply:
thanks, refined